File: sdks/sandbox/csharp/src/OpenSandbox/Internal/LifecycleMetricsReporter.cs
// Copyright 2026 Alibaba Group Holding Ltd.
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

using System.Net.Http.Headers;
using System.Text.Json;
using System.Text.Json.Serialization;
using Microsoft.Extensions.Logging;
using Microsoft.Extensions.Logging.Abstractions;
using OpenSandbox.Config;
using OpenSandbox.Core;

namespace OpenSandbox.Internal;

/// <summary>
/// Fire-and-forget reporter for SDK-side sandbox lifecycle metrics.
///
/// Behavior matches the other language SDKs (Go/JS/Python):
/// posts a small JSON body to <c>{baseUrl}/metrics/events</c> and never
/// blocks or surfaces errors to the caller. Honors the connection-level
/// <see cref="ConnectionConfig.DisableMetrics"/> opt-out and the
/// <c>OPENSANDBOX_DISABLE_METRICS=1</c> environment variable.
/// </summary>
internal static class LifecycleMetricsReporter
{
    private static readonly JsonSerializerOptions SerializerOptions = new()
    {
        DefaultIgnoreCondition = JsonIgnoreCondition.WhenWritingNull,
        Encoder = System.Text.Encodings.Web.JavaScriptEncoder.UnsafeRelaxedJsonEscaping,
    };

    /// <summary>
    /// Reports a sandbox.create latency event. Never throws.
    /// </summary>
    /// <param name="connectionConfig">Connection configuration used to derive URL, headers, and opt-out.</param>
    /// <param name="sandboxId">Optional created sandbox ID.</param>
    /// <param name="image">Optional image or snapshot reference used for creation.</param>
    /// <param name="createDurationMs">Measured create latency in milliseconds.</param>
    /// <param name="success">Whether the create flow succeeded end-to-end.</param>
    /// <param name="loggerFactory">Optional logger factory for debug-level failure logs.</param>
    public static void ReportSandboxCreate(
        ConnectionConfig connectionConfig,
        string? sandboxId,
        string? image,
        long createDurationMs,
        bool success,
        ILoggerFactory? loggerFactory = null)
    {
        if (connectionConfig.DisableMetrics)
        {
            return;
        }

        var logger = (loggerFactory ?? NullLoggerFactory.Instance)
            .CreateLogger("OpenSandbox.LifecycleMetricsReporter");

        // Telemetry is best-effort and MUST NOT surface any exception to the
        // caller. In particular, this method is invoked from Sandbox.CreateAsync's
        // failure path; any exception thrown while building the payload/URL would
        // otherwise replace the original create failure. Guard the whole method
        // body — including construction — with a top-level try/catch.
        byte[] body;
        string url;
        try
        {
            var payload = BuildPayload(sandboxId, image, createDurationMs, success);
            body = JsonSerializer.SerializeToUtf8Bytes(payload, SerializerOptions);
            url = $"{connectionConfig.GetBaseUrl().TrimEnd('/')}/metrics/events";
        }
        catch (Exception ex)
        {
            logger.LogDebug(ex, "Failed to build sandbox.create metrics request");
            return;
        }

        // Fire-and-forget: swallow any exception. Never block the caller.
        _ = Task.Run(async () =>
        {
            try
            {
                using var client = new HttpClient
                {
                    Timeout = TimeSpan.FromSeconds(Math.Max(1, connectionConfig.RequestTimeoutSeconds)),
                };
                using var content = new ByteArrayContent(body);
                content.Headers.ContentType = new MediaTypeHeaderValue("application/json");

                using var request = new HttpRequestMessage(HttpMethod.Post, url) { Content = content };
                request.Headers.UserAgent.Clear();
                request.Headers.UserAgent.ParseAdd(connectionConfig.UserAgent);
                foreach (var header in connectionConfig.Headers)
                {
                    // Skip content headers here; content-type is set above.
                    if (string.Equals(header.Key, "Content-Type", StringComparison.OrdinalIgnoreCase))
                    {
                        continue;
                    }
                    request.Headers.TryAddWithoutValidation(header.Key, header.Value);
                }
                if (!string.IsNullOrEmpty(connectionConfig.ApiKey))
                {
                    request.Headers.Remove(Constants.ApiKeyHeader);
                    request.Headers.TryAddWithoutValidation(Constants.ApiKeyHeader, connectionConfig.ApiKey);
                }

                using var response = await client.SendAsync(request).ConfigureAwait(false);
                // Drain content to allow connection reuse; ignore status.
                _ = await response.Content.ReadAsByteArrayAsync().ConfigureAwait(false);
            }
            catch (Exception ex)
            {
                logger.LogDebug(ex, "Failed to report sandbox.create metrics");
            }
        });
    }

    internal static IReadOnlyDictionary<string, object> BuildPayload(
        string? sandboxId,
        string? image,
        long createDurationMs,
        bool success)
    {
        var payload = new Dictionary<string, object>
        {
            ["eventType"] = "sandbox.create",
            ["createDurationMs"] = Math.Max(0, createDurationMs),
            ["success"] = success,
        };
        if (!string.IsNullOrEmpty(sandboxId))
        {
            payload["sandboxId"] = sandboxId!;
        }
        if (!string.IsNullOrEmpty(image))
        {
            payload["image"] = image!;
        }
        return payload;
    }
}

File: sdks/sandbox/kotlin/sandbox/src/main/kotlin/com/alibaba/opensandbox/sandbox/internal/LifecycleMetricsReporter.kt
/*
 * Copyright 2026 Alibaba Group Holding Ltd.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package com.alibaba.opensandbox.sandbox.internal

import com.alibaba.opensandbox.sandbox.config.ConnectionConfig
import kotlinx.serialization.json.JsonPrimitive
import kotlinx.serialization.json.buildJsonObject
import okhttp3.Call
import okhttp3.Callback
import okhttp3.MediaType.Companion.toMediaType
import okhttp3.OkHttpClient
import okhttp3.Request
import okhttp3.RequestBody.Companion.toRequestBody
import okhttp3.Response
import org.slf4j.LoggerFactory
import java.io.IOException
import java.util.concurrent.TimeUnit

/**
 * Fire-and-forget reporter for SDK-side sandbox lifecycle metrics.
 *
 * Behavior matches the other language SDKs (Go/JS/Python/C#):
 * posts a small JSON body to `{baseUrl}/metrics/events` and never blocks
 * or surfaces errors to the caller. Honors both the connection-level
 * [ConnectionConfig.disableMetrics] opt-out and the
 * `OPENSANDBOX_DISABLE_METRICS=1` environment variable.
 */
internal object LifecycleMetricsReporter {
    private val logger = LoggerFactory.getLogger(LifecycleMetricsReporter::class.java)
    private val jsonMediaType = "application/json".toMediaType()

    /**
     * Best-effort reports the sandbox.create latency event. Never throws.
     *
     * @param connectionConfig connection configuration used to derive URL, headers, and opt-out
     * @param sandboxId optional created sandbox ID (omitted from payload when null)
     * @param image optional image or snapshot reference used for creation (omitted when null)
     * @param createDurationMs measured create latency in milliseconds (clamped to >= 0)
     * @param success whether the create flow succeeded end-to-end
     * @param client optional OkHttp client override, primarily for tests
     */
    @JvmStatic
    @JvmOverloads
    fun reportSandboxCreate(
        connectionConfig: ConnectionConfig,
        sandboxId: String?,
        image: String?,
        createDurationMs: Long,
        success: Boolean,
        client: OkHttpClient? = null,
    ) {
        if (connectionConfig.isCreateMetricsDisabled()) {
            return
        }

        // Telemetry is best-effort and MUST NOT surface any exception to the caller.
        // In particular, this method is invoked from the Sandbox.create failure path;
        // any exception thrown from payload/URL/request construction (e.g. an invalid
        // configured baseUrl making Request.Builder.url(...) throw
        // IllegalArgumentException) would otherwise replace the original create
        // failure. Guard the whole body — not just the enqueue call — so every
        // telemetry failure is swallowed.
        try {
            val payload = buildPayload(sandboxId, image, createDurationMs, success)
            val url = connectionConfig.getBaseUrl().trimEnd('/') + "/metrics/events"

            val requestBuilder =
                Request.Builder()
                    .url(url)
                    .post(payload.toRequestBody(jsonMediaType))
                    .header("Content-Type", "application/json")
                    .header("User-Agent", connectionConfig.userAgent)

            val apiKey = connectionConfig.getApiKey()
            if (apiKey.isNotBlank()) {
                requestBuilder.header("OPEN-SANDBOX-API-KEY", apiKey)
            }
            for ((name, value) in connectionConfig.headers) {
                // Do not override the auth header if the user provided one via headers.
                if (name.equals("OPEN-SANDBOX-API-KEY", ignoreCase = true) ||
                    name.equals("Content-Type", ignoreCase = true) ||
                    name.equals("User-Agent", ignoreCase = true)
                ) {
                    requestBuilder.header(name, value)
                } else {
                    requestBuilder.addHeader(name, value)
                }
            }

            // When the caller supplies an OkHttpClient, they own its lifecycle. Otherwise
            // we create a client for this single fire-and-forget request and tear it down
            // in the callback so that high-churn create paths (e.g. SandboxPool prewarm)
            // do not accumulate OkHttp dispatcher threads or idle connections.
            val effectiveClient: OkHttpClient
            val sdkOwnsClient: Boolean
            if (client != null) {
                effectiveClient = client
                sdkOwnsClient = false
            } else {
                effectiveClient = defaultClient(connectionConfig)
                sdkOwnsClient = true
            }

            try {
                effectiveClient.newCall(requestBuilder.build()).enqueue(
                    object : Callback {
                        override fun onFailure(
                            call: Call,
                            e: IOException,
                        ) {
                            try {
                                logger.debug("Failed to report sandbox.create metrics: {}", e.message)
                            } finally {
                                if (sdkOwnsClient) shutdownQuietly(effectiveClient)
                            }
                        }

                        override fun onResponse(
                            call: Call,
                            response: Response,
                        ) {
                            try {
                                // Drain and close to release the connection back to the pool.
                                response.use { it.body?.string() }
                            } catch (_: Exception) {
                                // ignore — best effort
                            } finally {
                                if (sdkOwnsClient) shutdownQuietly(effectiveClient)
                            }
                        }
                    },
                )
            } catch (e: Exception) {
                // Never let telemetry disrupt the caller.
                logger.debug("Failed to enqueue sandbox.create metrics request: {}", e.message)
                if (sdkOwnsClient) shutdownQuietly(effectiveClient)
            }
        } catch (e: Exception) {
            // Payload/URL/request construction failed (e.g. invalid baseUrl). Swallow.
            logger.debug("Failed to build sandbox.create metrics request: {}", e.message)
        }
    }

    /**
     * Shuts down an SDK-owned OkHttpClient created solely for a single
     * telemetry request. Never throws.
     */
    private fun shutdownQuietly(client: OkHttpClient) {
        try {
            client.dispatcher.executorService.shutdown()
        } catch (_: Exception) {
            // ignore
        }
        try {
            client.connectionPool.evictAll()
        } catch (_: Exception) {
            // ignore
        }
    }

    /**
     * Builds a compact JSON payload for the sandbox.create metrics event.
     * `sandboxId` and `image` are omitted when null/blank to match the wire
     * format used by the other SDKs.
     */
    internal fun buildPayload(
        sandboxId: String?,
        image: String?,
        createDurationMs: Long,
        success: Boolean,
    ): String {
        val obj =
            buildJsonObject {
                put("eventType", JsonPrimitive("sandbox.create"))
                put("createDurationMs", JsonPrimitive(maxOf(0L, createDurationMs)))
                put("success", JsonPrimitive(success))
                if (!sandboxId.isNullOrBlank()) {
                    put("sandboxId", JsonPrimitive(sandboxId))
                }
                if (!image.isNullOrBlank()) {
                    put("image", JsonPrimitive(image))
                }
            }
        return obj.toString()
    }

    private fun defaultClient(connectionConfig: ConnectionConfig): OkHttpClient {
        val timeoutMs = maxOf(1L, connectionConfig.requestTimeout.toMillis())
        return OkHttpClient.Builder()
            .connectTimeout(timeoutMs, TimeUnit.MILLISECONDS)
            .readTimeout(timeoutMs, TimeUnit.MILLISECONDS)
            .writeTimeout(timeoutMs, TimeUnit.MILLISECONDS)
            .callTimeout(timeoutMs, TimeUnit.MILLISECONDS)
            .build()
    }
}
